In [224]:
# Imports and pip installations (if needed)
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import MinMaxScaler
import numpy as np

# Part 1: Load the dataset

In [225]:
# Load the given datasets
ckd_categorical = pd.read_csv("chronic_kidney_disease_categorical.csv")
ckd_numerical = pd.read_csv("chronic_kidney_disease_numerical.csv")


# Print the data
print(ckd_categorical)
print(ckd_numerical)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good  yes   no     c

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:

unique_id: A unique identifier for each patient record (not a feature for prediction).

al: Albumin level in urine (categorical - typically indicates the presence of albumin).

su: Sugar level in urine (categorical - indicates whether sugar is present).

rbc: Presence of red blood cells in urine (categorical - can be "normal" or "abnormal").

pc: Presence of pus cells in urine (categorical - can be "normal" or "abnormal").

pcc: Presence of pus cell clumps (categorical - indicates presence or absence).

ba: Presence of bacteria (categorical - indicates presence or absence).

htn: Hypertension status (binary categorical - "yes" or "no").

dm: Diabetes mellitus status (binary categorical - "yes" or "no").

cad: Coronary artery disease status (binary categorical - "yes" or "no").

appet: Appetite status (binary categorical - "good" or "poor").

pe: Pedal edema status (binary categorical - "yes" or "no").

ane: Anemia status (binary categorical - "yes" or "no").

Target: Indication of chronic kidney disease status (binary categorical - "ckd" or "notckd").

bu: Blood urea level (numerical - measured in mg/dl).

sc: Serum creatinine level (numerical - measured in mg/dl).

sod: Sodium level in the blood (numerical - measured in mEq/L).

pot: Potassium level in the blood (numerical - measured in mEq/L).

hemo: Hemoglobin level (numerical - measured in grams).

pcv: Packed cell volume (numerical - typically a percentage).

wbcc: White blood cell count (numerical - measured in cells/cmm).

rbcc: Red blood cell count (numerical - measured in millions/cmm).

age: Age of the patient (numerical - measured in years).

bp: Blood pressure (numerical - measured in mm/Hg).

sg: Specific gravity (categorical - indicates concentration).

bgr: Blood glucose random level (numerical - measured in mg/dl).


Features: The independent variables used for prediction. In this dataset, features include:
`al`, `su`, `rbc`, `pc`, `pcc`, `ba`, `htn`, `dm`, `cad`, `appet`, `pe`, `ane`, `bu`, `sc`, `sod`, `pot`, `hemo`, `pcv`, `wbcc`, `rbcc`, `age`, `bp`, `sg`, `bgr`.

Label: The output variable that we want to predict, which is:
`Target`: Indicates whether the patient has chronic kidney disease (CKD) or not (values: "ckd" or "notckd").


Categorical Features Examples:
1. `rbc` (presence of red blood cells: "normal" or "abnormal")
2. `htn` (hypertension: "yes" or "no")
3. `dm` (diabetes: "yes" or "no")

Numerical Features Examples:
1. `bu` (blood urea level in mg/dl)
2. `sc` (serum creatinine level in mg/dl)
3. `age` (age of the patient in years)

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [226]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
duplicates_numerical = ckd_numerical.duplicated()
num_duplicates_numerical = duplicates_numerical.sum()
print(f"No. of duplicate rows in ckd_numerical: {num_duplicates_numerical}")

# Drop these duplicate rows
ckd_numerical = ckd_numerical.drop_duplicates()


# Repeat the same for categorical dataset. Print the duplicate rows and drop them
duplicates_categorical = ckd_categorical.duplicated()
num_duplicates_categorical = duplicates_categorical.sum()
print(f"No. of duplicate rows in ckd_categorical: {num_duplicates_categorical}")

ckd_categorical = ckd_categorical.drop_duplicates()




No. of duplicate rows in ckd_numerical: 30
No. of duplicate rows in ckd_categorical: 30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: To determine if all unique IDs are present in both datasets, we can check for any null or NaN values in the columns 'ckd_categorical' and 'ckd_numerical' before merging. While each row in both datasets has a unique identifier, the 'ckd_categorical' dataset contains 400 rows, and the 'ckd_numerical' dataset has 390 rows. This indicates that there are 10 extra rows in the categorical dataset. When merging the two datasets using an outer join, these 10 additional rows from 'ckd_categorical' will appear with NaN values for the corresponding missing data from 'ckd_numerical.'

In [227]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
print(ckd_categorical.shape)
print(ckd_numerical.shape)
print(ckd_categorical['unique_id'].isnull().sum())
print(ckd_numerical['unique_id'].isnull().sum())

ckd_merged = pd.merge(ckd_categorical, ckd_numerical, on='unique_id', how='outer')

print(ckd_merged.shape)
print(ckd_merged['unique_id'].isnull().sum())

#Print the combined dataset
print(ckd_merged)

(400, 14)
(390, 12)
0
0
(400, 25)
0
     unique_id   al   su     rbc        pc         pcc          ba  htn   dm  \
0       102677  3.0  0.0  normal  abnormal  notpresent  notpresent  yes   no   
1       103570  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
2       104675  2.0  1.0  normal    normal     present  notpresent   no  yes   
3       105923  1.0  0.0     NaN    normal  notpresent  notpresent  yes  yes   
4       107297  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
..         ...  ...  ...     ...       ...         ...         ...  ...  ...   
395     995177  4.0  3.0  normal  abnormal     present     present  yes  yes   
396     996076  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
397     996412  NaN  NaN     NaN       NaN  notpresent  notpresent   no   no   
398     997629  0.0  0.0  normal    normal         NaN         NaN   no   no   
399     997876  0.0  0.0     NaN    normal  notpresent  notpresent   no   no   

   

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [228]:
# Calculate the percentage of rows that contain atleast one missing value
missing_values = ckd_merged.isnull().any(axis=1)

# Print %
percent_missing = (missing_values.sum() / len(ckd_merged)) * 100
print(f"Percentage of rows that contain atleast one missing value: {percent_missing}%")

# Drop these rows from the dataset
ckd_merged.dropna(inplace=True)

# Print the Dataset
print(ckd_merged)


Percentage of rows that contain atleast one missing value: 61.75000000000001%
     unique_id   al   su     rbc        pc         pcc          ba  htn   dm  \
0       102677  3.0  0.0  normal  abnormal  notpresent  notpresent  yes   no   
6       109053  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
7       114717  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
8       118191  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
9       118596  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
..         ...  ...  ...     ...       ...         ...         ...  ...  ...   
390     980291  3.0  0.0  normal  abnormal  notpresent  notpresent  yes  yes   
391     991031  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
394     994377  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
395     995177  4.0  3.0  normal  abnormal     present     present  yes  yes   
396     996076  0.0  0.0  normal    normal

## Part 3.4: Sort the dataset according to the Labels

In [229]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
ckd_merged.sort_values(by='Target', inplace=True)
ckd_merged.reset_index(drop=True, inplace=True)

# Print the dataset
print(ckd_merged)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       102677  3.0  0.0    normal  abnormal  notpresent  notpresent  yes   
1       337834  4.0  0.0    normal  abnormal     present  notpresent  yes   
2       343710  2.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       349892  1.0  3.0  abnormal  abnormal  notpresent  notpresent  yes   
4       353025  3.0  4.0    normal    normal  notpresent  notpresent  yes   
..         ...  ...  ...       ...       ...         ...         ...  ...   
148     486397  0.0  0.0    normal    normal  notpresent  notpresent   no   
149     203466  0.0  0.0    normal    normal  notpresent  notpresent   no   
150     482105  0.0  0.0    normal    normal  notpresent  notpresent   no   
151     228849  0.0  0.0    normal    normal  notpresent  notpresent   no   
152     996076  0.0  0.0    normal    normal  notpresent  notpresent   no   

      dm  cad  ...    bp    bgr     bu    sc    sod  pot  hemo   pcv     wb

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [230]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']

ckd_merged = pd.get_dummies(ckd_merged, columns=categorical_columns, dtype='int')

# Print the dataset
print(ckd_merged)

     unique_id   al   su   age    bp    bgr     bu    sc    sod  pot  ...  \
0       102677  3.0  0.0  59.0  70.0   76.0  186.0  15.0  135.0  7.6  ...   
1       337834  4.0  0.0  48.0  70.0  117.0   56.0   3.8  111.0  2.5  ...   
2       343710  2.0  0.0  61.0  80.0  173.0  148.0   3.9  135.0  5.2  ...   
3       349892  1.0  3.0  59.0  70.0  424.0   55.0   1.7  138.0  4.5  ...   
4       353025  3.0  4.0  40.0  70.0  253.0  150.0  11.9  132.0  5.6  ...   
..         ...  ...  ...   ...   ...    ...    ...   ...    ...  ...  ...   
148     486397  0.0  0.0  20.0  70.0  123.0   44.0   1.0  135.0  3.8  ...   
149     203466  0.0  0.0  22.0  60.0   97.0   18.0   1.2  138.0  4.3  ...   
150     482105  0.0  0.0  44.0  60.0   95.0   46.0   0.5  138.0  4.2  ...   
151     228849  0.0  0.0  58.0  80.0  100.0   50.0   1.2  140.0  3.5  ...   
152     996076  0.0  0.0  57.0  60.0  105.0   49.0   1.2  150.0  4.7  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_ye

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: Using a single column works for binary variables but doesn't for categories with more than two values. In your "type of condiment" example with 5 options, you'd need to use one-hot encoding (one column per category). A single column wouldn't capture all the necessary information, and the model's performance could suffer.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [231]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
ckd_merged_copy = ckd_merged.copy()

for column in numerical_columns:
    mean = ckd_merged[column].mean() 
    std = ckd_merged[column].std()  
    
    lower_bound = mean - 3 * std
    upper_bound = mean + 3 * std
    
    ckd_merged_copy = ckd_merged_copy[(ckd_merged_copy[column] >= lower_bound) & (ckd_merged_copy[column] <= upper_bound)]

ckd_merged = ckd_merged_copy.reset_index(drop='True')
                                    

# Print the dataset
print(ckd_merged)

     unique_id   al   su   age     bp    bgr     bu   sc    sod  pot  ...  \
0       343710  2.0  0.0  61.0   80.0  173.0  148.0  3.9  135.0  5.2  ...   
1       397388  2.0  2.0  63.0  100.0  280.0   35.0  3.2  143.0  3.5  ...   
2       438182  2.0  0.0  73.0   80.0  253.0  142.0  4.6  138.0  5.8  ...   
3       474407  4.0  0.0  21.0   90.0  107.0   40.0  1.7  125.0  3.5  ...   
4       475200  4.0  2.0  64.0  100.0  163.0   54.0  7.2  140.0  4.6  ...   
..         ...  ...  ...   ...    ...    ...    ...  ...    ...  ...  ...   
129     486397  0.0  0.0  20.0   70.0  123.0   44.0  1.0  135.0  3.8  ...   
130     203466  0.0  0.0  22.0   60.0   97.0   18.0  1.2  138.0  4.3  ...   
131     482105  0.0  0.0  44.0   60.0   95.0   46.0  0.5  138.0  4.2  ...   
132     228849  0.0  0.0  58.0   80.0  100.0   50.0  1.2  140.0  3.5  ...   
133     996076  0.0  0.0  57.0   60.0  105.0   49.0  1.2  150.0  4.7  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_ye

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [232]:
# Normalize the all Numerical Attributes in the dataset.
scaler = MinMaxScaler()
ckd_merged[numerical_columns] = scaler.fit_transform(ckd_merged[numerical_columns])


# Print the dataset
print(ckd_merged)

     unique_id   al   su       age    bp       bgr        bu        sc  \
0       343710  2.0  0.0  0.743243  0.50  0.442060  0.901961  0.432099   
1       397388  2.0  2.0  0.770270  1.00  0.901288  0.163399  0.345679   
2       438182  2.0  0.0  0.905405  0.50  0.785408  0.862745  0.518519   
3       474407  4.0  0.0  0.202703  0.75  0.158798  0.196078  0.160494   
4       475200  4.0  2.0  0.783784  1.00  0.399142  0.287582  0.839506   
..         ...  ...  ...       ...   ...       ...       ...       ...   
129     486397  0.0  0.0  0.189189  0.25  0.227468  0.222222  0.074074   
130     203466  0.0  0.0  0.216216  0.00  0.115880  0.052288  0.098765   
131     482105  0.0  0.0  0.513514  0.00  0.107296  0.235294  0.012346   
132     228849  0.0  0.0  0.702703  0.50  0.128755  0.261438  0.098765   
133     996076  0.0  0.0  0.689189  0.00  0.150215  0.254902  0.098765   

          sod       pot  ...  cad_no  cad_yes  appet_good  appet_poor  pe_no  \
0    0.500000  0.657143  ...   

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: This column is simply a unique identifier for each patient and does not contain any useful information for predicting chronic kidney disease (CKD). It does not have any relationship with the target variable and can be safely removed.

In [233]:
#Remove that column
ckd_merged.drop(columns=['unique_id'], inplace=True)


# Print the dataset
print(ckd_merged.head(15))

     al   su       age    bp       bgr        bu        sc       sod  \
0   2.0  0.0  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
1   2.0  2.0  0.770270  1.00  0.901288  0.163399  0.345679  0.766667   
2   2.0  0.0  0.905405  0.50  0.785408  0.862745  0.518519  0.600000   
3   4.0  0.0  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
4   4.0  2.0  0.783784  1.00  0.399142  0.287582  0.839506  0.666667   
5   3.0  1.0  0.729730  0.00  0.935622  0.169935  0.160494  0.333333   
6   2.0  0.0  0.675676  0.75  0.253219  0.633987  0.777778  0.366667   
7   4.0  3.0  0.851351  0.25  0.618026  0.562092  0.728395  0.000000   
8   1.0  0.0  0.540541  0.00  0.399142  0.535948  0.358025  0.700000   
9   3.0  0.0  0.756757  0.25  0.223176  0.209150  0.160494  0.533333   
10  3.0  1.0  0.662162  0.50  0.618026  0.411765  0.432099  0.566667   
11  4.0  1.0  0.783784  0.00  0.725322  0.313725  0.481481  0.566667   
12  4.0  0.0  0.878378  0.00  0.206009  0.751634  0.604938  0.53

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [234]:
# Export the dataframe to a new csv file
ckd_merged.to_csv('cleaned_ckd_dataset.csv', index=False)

# Export the dataframe to a new json file
ckd_merged.to_json('cleaned_ckd_dataset.json', orient='records', lines=True)

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

|    | al   | su   | age      | bp   | bgr      | bu       | sc       | sod      | pot      | hemo     | ... | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|----|------|------|----------|------|----------|----------|----------|----------|----------|----------|-----|--------|---------|------------|-------------|-------|--------|--------|---------|-------------|----------------|
| 0  | 2.0  | 0.0  | 0.743243 | 0.50 | 0.442060 | 0.901961 | 0.432099 | 0.500000 | 0.657143 | 0.172131 | ... | 0      | 1       | 0          | 1           | 0     | 1      | 0      | 1       | 1           | 0              |
| 1  | 2.0  | 2.0  | 0.770270 | 1.00 | 0.901288 | 0.163399 | 0.345679 | 0.766667 | 0.171429 | 0.606557 | ... | 0      | 1       | 1          | 0           | 1     | 0      | 1      | 0       | 1           | 0              |
| 2  | 2.0  | 0.0  | 0.905405 | 0.50 | 0.785408 | 0.862745 | 0.518519 | 0.600000 | 0.828571 | 0.401639 | ... | 0      | 1       | 1          | 0           | 1     | 0      | 1      | 0       | 1           | 0              |
| 3  | 4.0  | 0.0  | 0.202703 | 0.75 | 0.158798 | 0.196078 | 0.160494 | 0.166667 | 0.171429 | 0.221311 | ... | 1      | 0       | 1          | 0           | 1     | 0      | 0      | 1       | 1           | 0              |
| 4  | 4.0  | 2.0  | 0.783784 | 1.00 | 0.399142 | 0.287582 | 0.839506 | 0.666667 | 0.485714 | 0.188525 | ... | 1      | 0       | 1          | 0           | 0     | 0      | 1      | 0       | 1           | 0              |
| 5  | 3.0  | 1.0  | 0.729730 | 0.00 | 0.935622 | 0.169935 | 0.160494 | 0.333333 | 0.028571 | 0.188525 | ... | 1      | 0       | 0          | 1           | 1     | 0      | 0      | 1       | 1           | 0              |
| 6  | 2.0  | 0.0  | 0.675676 | 0.75 | 0.253219 | 0.633987 | 0.777778 | 0.366667 | 0.542857 | 0.286885 | ... | 1      | 0       | 1          | 0           | 1     | 0      | 1      | 0       | 1           | 0              |
| 7  | 4.0  | 3.0  | 0.851351 | 0.25 | 0.618026 | 0.562092 | 0.728395 | 0.000000 | 0.285714 | 0.311475 | ... | 0      | 1       | 1          | 0           | 0     | 1      | 0      | 1       | 1           | 0              |
| 8  | 1.0  | 0.0  | 0.540541 | 0.00 | 0.399142 | 0.535948 | 0.358025 | 0.700000 | 0.314286 | 0.344262 | ... | 1      | 0       | 1          | 0           | 1     | 0      | 1      | 0       | 1           | 0              |
| 9  | 3.0  | 0.0  | 0.756757 | 0.25 | 0.223176 | 0.209150 | 0.160494 | 0.533333 | 0.514286 | 0.573770 | ... | 1      | 0       | 1          | 0           | 1     | 0      | 1      | 0       | 1           | 0              |
| 10 | 3.0  | 1.0  | 0.662162 | 0.50 | 0.618026 | 0.411765 | 0.432099 | 0.566667 | 0.571429 | 0.434426 | ... | 1      | 0       | 1          | 0           | 0     | 1      | 1      | 0       | 1           | 0              |
| 11 | 4.0  | 1.0  | 0.783784 | 0.00 | 0.725322 | 0.313725 | 0.481481 | 0.566667 | 0.714286 | 0.319672 | ... | 1      | 0       | 0          | 1           | 0     | 1      | 1      | 0       | 1           | 0              |
| 12 | 4.0  | 0.0  | 0.878378 | 0.00 | 0.206009 | 0.751634 | 0.604938 | 0.533333 | 0.571429 | 0.475410 | ... | 1      | 0       | 0          | 1           | 0     | 1      | 1      | 0       | 1           | 0              |
| 13 | 3.0  | 0.0  | 0.878378 | 0.25 | 0.639485 | 0.470588 | 0.395062 | 0.433333 | 0.428571 | 0.393443 | ... | 0      | 1       | 1          | 0           | 1     | 0      | 1      | 0       | 1           | 0              |
| 14 | 1.0  | 0.0  | 0.716216 | 0.50 | 1.000000 | 0.163399 | 0.111111 | 0.066667 | 0.171429 | 0.393443 | ... | 1      | 0       | 0          | 1           | 0     | 0      | 1      | 0       | 1           | 0              |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

To mitigate data conversion hallucinations in large language models (LLMs), it's essential to ensure high-quality training data and consider fine-tuning the model on relevant domain-specific content. Effective prompt engineering through clear, specific instructions and providing contextual information is crucial to guide the model appropriately. Implementing verification mechanisms, such as cross-checking outputs against reputable sources and incorporating human oversight for critical tasks, can further ensure accuracy. Additionally, utilizing feedback from users to retrain the model and possibly employing ensemble methods with multiple models can help identify and correct inaccuracies, ultimately creating a more reliable output.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [235]:
# Code to rename all the columns in the dataset
rename_dict = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class'
}

ckd_merged.rename(columns=rename_dict, inplace=True)

print(ckd_merged.head(15))

    albumin  sugar       age  blood pressure  blood glucose random  \
0       2.0    0.0  0.743243            0.50              0.442060   
1       2.0    2.0  0.770270            1.00              0.901288   
2       2.0    0.0  0.905405            0.50              0.785408   
3       4.0    0.0  0.202703            0.75              0.158798   
4       4.0    2.0  0.783784            1.00              0.399142   
5       3.0    1.0  0.729730            0.00              0.935622   
6       2.0    0.0  0.675676            0.75              0.253219   
7       4.0    3.0  0.851351            0.25              0.618026   
8       1.0    0.0  0.540541            0.00              0.399142   
9       3.0    0.0  0.756757            0.25              0.223176   
10      3.0    1.0  0.662162            0.50              0.618026   
11      4.0    1.0  0.783784            0.00              0.725322   
12      4.0    0.0  0.878378            0.00              0.206009   
13      3.0    0.0  

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [236]:
# Converted SQL to Pandas code
ckd_merged['Target'] = ckd_merged['Target_ckd'].replace({1: 'ckd', 0:'not_ckd'})
result = ckd_merged.groupby('Target').size().reset_index(name='count')

print(result)


    Target  count
0      ckd     23
1  not_ckd    111


*Describe what the above code does here*

The code above uses the Pandas library to group a DataFrame (`ckd_merged`) by the 'Target' column and count the occurrences of each unique value in that column. It performs a group-by operation with `df.groupby('Target')`, then counts the rows in each group using `.size()`. The results are converted back into a DataFrame with `reset_index(name='count')`, creating two columns: 'Target' and 'count', which represent the unique values in 'Target' and their corresponding counts, respectively. Finally, the code prints this resulting DataFrame, effectively mirroring the logic of the provided SQL query.